src/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs

```diff
@@ -111,7 +114,42 @@ public string GetHeaderValue(ReadOnlySpan<byte> headerValue)
                 }
             }
 
+            if (KnownHeader != null && KnownHeader.Name == KnownHeaders.Location.Name)
+            {
+                // Normally Location should be in ISO-8859-1, ocassionally some servers respond with UTF-8 though
+                if (TryDecodeUtf8(headerValue, out string decoded))
+                {
+                    return decoded;
+                }
+            }
+
             return HttpRuleParser.DefaultHttpEncoding.GetString(headerValue);
         }
+
+        private static bool TryDecodeUtf8(ReadOnlySpan<byte> input, out string decoded)
+        {
+            bool possibleUtf8 = false;
+            for (int i = 0; i < input.Length; i++)
+            {
+                if (input[i] > 127)
+                {
+                    possibleUtf8 = true;
+                    break;
+                }
+            }
+
+            if (possibleUtf8)
+            {
+                try
+                {
+                    decoded = s_utf8DecoderWithExceptionFallback.GetString(input);
+                    return true;
+                }
+                catch (ArgumentException) { } // Not actually Utf-8
```

Comment:
> catch (ArgumentException) { } // Not actually Utf-8 [](start = 15, length = 52)

nit: if you want to avoid throwing and catch, maybe you can just search for invalid character or provide your own fallback to detect that.

Reply:
@tarekgh Seems I'd have to create a fallback instance per conversion to do that, right? Isn't that going to be slower for the normal case?

Reply:
I think I'll leave this as is - perhaps @GrabYourPitchforks has some utility to do the conversion/validation without throwing.

Reply:
Sounds good.